### PR TITLE
[Arm64] Don't use D-copies in CopyBlock

### LINKED ISSUE #63453: [Perf] Changes at 12/28/2021 6:18:01 PM
### Run Information

Architecture | arm64
-- | --
OS | Windows 10.0.19041
Baseline | [d094a34a4d555d3aef434bc87ad3fefe15d83e08](https://github.com/dotnet/runtime/commit/d094a34a4d555d3aef434bc87ad3fefe15d83e08)
Compare | [15e95fc44be0ea444ac145b3cc2d4207fa1a6fef](https://github.com/dotnet/runtime/commit/15e95fc44be0ea444ac145b3cc2d4207fa1a6fef)
Diff | [Diff](https://github.com/dotnet/runtime/compare/d094a34a4d555d3aef434bc87ad3fefe15d83e08...15e95fc44be0ea444ac145b3cc2d4207fa1a6fef)

### Regressions in System.Numerics.Tests.Perf_Matrix4x4

Benchmark | Baseline | Test | Test/Base | Test Quality | Edge Detector | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
[InequalityOperatorBenchmark - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Numerics.Tests.Perf_Matrix4x4.InequalityOperatorBenchmark.html>) | 11.74 ns | 13.11 ns | 1.12 | 0.12 | False | | |

graph
[Test Report](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/01_05_2022/refs/heads/main_arm64_Windows%2010.0.19041_Regression/System.Numerics.Tests.Perf_Matrix4x4.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f net6.0 --filter 'System.Numerics.Tests.Perf_Matrix4x4*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-5ea41fda-1585-4888-b701-85db863cc8dd779557eaf20483896/0c5885dc-c73d-465d-9c39-b61e13fa89d6.zip?sv=2019-07-07&se=2022-01-27T06%3A17%3A35Z&sr=c&sp=rl&sig=sZVX%2Buyn%2B9GwWa7RUTSrU2UC0ErgCCWdwpZdB1Z5bpE%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-d65b0eb2-c23b-49fc-a960-15752c7d46bf8053481513a41b9a1/2a3de55e-6258-4ba8-807d-7e3f65e5d9d9.zip?sv=2019-07-07&se=2022-01-28T00%3A50%3A21Z&sr=c&sp=rl&sig=i%2BFxkbYhujzbSsLzLSzzsk4ribRLXNSKGVWpF91Svo4%3D>)
#### System.Numerics.Tests.Perf_Matrix4x4.InequalityOperatorBenchmark

```log

```
### Description of detection logic
```IsRegressionBase: Marked as regression because the compare was 5% greater than the baseline, and the value was not too small.
IsRegressionChecked: Marked as regression because the three check build points were 0.05 greater than the baseline.
IsImprovementBase: Marked as not an improvement because the compare was not 5% less than the baseline, or the value was too small.
IsRegressionBase: Marked as regression because the compare was 5% greater than the baseline, and the value was not too small.
IsRegressionChecked: Marked as regression because the three check build points were 0.05 greater than the baseline.
IsRegressionWindowed: Marked as regression because 13.109460078459284 > 12.288147412937622.
IsChangePoint: Marked as a change because one of 12/28/2021 8:42:38 AM, 1/5/2022 11:37:43 PM falls between 12/28/2021 8:42:38 AM and 1/5/2022 11:37:43 PM.
IsRegressionStdDev: Marked as regression because -19.598191341561044 (T) = (0 -13.018293385362945) / Math.Sqrt((0.01873680514312647 / (8)) + (0.06181210359770353 / (25))) is less than -2.0395134463927462 = MathNet.Numerics.Distributions.StudentT.InvCDF(0, 1, (8) + (25) - 2, .025) and -0.11664210781814657 = (11.658429584748447 - 13.018293385362945) / 11.658429584748447 is less than -0.05.
IsImprovementBase: Marked as not an improvement because the compare was not 5% less than the baseline, or the value was too small.
IsChangeEdgeDetector: Marked not as a regression because Edge Detector said so.

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by echesakov**
Resolves #63453:

|                      Method |        Job |                   Toolchain |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | Allocated |
|---------------------------- |----------- |---------------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|----------:|
| InequalityOperatorBenchmark | Job-YASJMH | \base\Core_Root\corerun.exe | 11.97 ns | 0.051 ns | 0.043 ns | 11.97 ns | 11.91 ns | 12.07 ns |  1.00 |         - |
| InequalityOperatorBenchmark | Job-RUKPGX | \diff\Core_Root\corerun.exe | 12.24 ns | 0.030 ns | 0.024 ns | 12.24 ns | 12.20 ns | 12.28 ns |  1.02 |         - |

Based on the discussion in https://github.com/dotnet/runtime/issues/63453#issuecomment-1007861814 don't use D-variants ldp/stp for copying 16 bytes and use integer ldp/stp (as before #61030). 

In order to allow this, we would need to allocate an additional integer register even if LSRA has allocated a pair of SIMD registers (the decision whether Q-copies can be used are based on whether both source and destination correspond to locals **and** their base addresses have the same `(startOffset % 16)` values, meaning that at LSRA phase we cannot determine whether the latter condition will hold since assignment of the addresses on the stack to locals is done later - at CodeGen phase).

@dotnet/jit-contrib PTAL
cc @TamarChristinaArm

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@dotnet/jit-contrib ping

**Comment by kunalspathak** [comment]
I will take a look later today.

**Comment by kunalspathak** [comment]
The changes looks good overall, but let's wait for @BruceForstall to have another look.

File: src/coreclr/jit/codegenarmarch.cpp [hunk 1]
```diff
@@ -2853,7 +2853,7 @@ void CodeGen::genCodeForCpBlkUnroll(GenTreeBlk* node)
     // be profitable).
 
     const bool canUse16ByteWideInstrs = isSrcRegAddrAlignmentKnown && isDstRegAddrAlignmentKnown &&
-                                        (size >= FP_REGSIZE_BYTES) && (srcRegAddrAlignment == dstRegAddrAlignment);
+                                        (size >= 2 * FP_REGSIZE_BYTES) && (srcRegAddrAlignment == dstRegAddrAlignment);
```
Comment (BruceForstall) [inline src/coreclr/jit/codegenarmarch.cpp:2856]: Does the comment above need to be adjusted? It does mention the D registers.

**Review by BruceForstall — APPROVED** [review]

File: src/coreclr/jit/codegenarmarch.cpp (hunk 1 above)
Comment (echesakovMSFT) [reply src/coreclr/jit/codegenarmarch.cpp:2856]: Yes, good catch, thanks. Updated.

**Pull request merged**